spec/support/matchers/accessibility.rb

```diff
@@ -109,6 +109,108 @@ def descriptors(element)
+  end
+
+  def aria_labelledby_name(element)
+    # "if computing a name, and the current node has an aria-labelledby attribute that contains at
+    # least one valid IDREF, and the current node is not already part of an aria-labelledby
+    # traversal, process its IDREFs in the order they occur"
+    valid_labels = element['aria-labelledby']&.
+      split(' ')&.
+      map { |label_id| page.find("##{label_id}")&.text }&.
+      compact
+
+    valid_labels.join('') if valid_labels.present?
+  end
+
+  def aria_label_name(element)
+    # "Otherwise, if computing a name, and if the current node has an aria-label attribute whose
+    # value is not the empty string, nor, when trimmed of white space, is not the empty string:"
+    element['aria-label']
+  end
+
+  def referenced_label_name(element)
+    # "Otherwise, if the current node's native markup provides an attribute (e.g. title) or element
+    # (e.g. HTML label) that defines a text alternative, return that alternative in the form of a
+    # flat string as defined by the host language"
+    descendent_name(page.find("label[for='#{element['id']}']"))
+  rescue Capybara::ElementNotFound
+    nil
+  end
+
+  def ancestor_label_text_name(element)
+    # "Otherwise, if the current node is a control embedded within the label"
+    descendent_name(element.find(:xpath, 'ancestor::label'))
+  rescue Capybara::ElementNotFound
+    nil
+  end
+
+  def descendent_name(element)
+    # "Otherwise, if the current node is a descendant of an element whose Accessible Name or
+    # Accessible Description is being computed, and contains descendants, proceed to 2F.i."
+    #
+    # Note: There should probably be some recursion here with the other methods, but Capybara only
+    # allows us to work with elements and not text nodes, which presents a challenge for computing
+    # the descendent name.
+    visible_text_js = <<~JS
+      (function getVisibleChildNodeText(childNodes) {
+          return Array.from(childNodes).flatMap((child) => {
+              switch (child.nodeType) {
+                  case Node.TEXT_NODE:
+                      return child.nodeValue;
+                  case Node.ELEMENT_NODE:
+                      if (child.getAttribute('aria-hidden') !== 'true') {
+                        return child.nodeName === 'IMG' ?
+                          child.getAttribute('alt') :
+                          getVisibleChildNodeText(child.childNodes);
+                      }
+              }
+          });
+      })(arguments[0].childNodes).filter(Boolean).join('').trim();
+    JS
+
+    page.evaluate_script(visible_text_js, element)
+  end
+
+  def computed_name(element)
+    Enumerator.new do |yielder|
+      yielder << hidden_name(element)
+      yielder << aria_labelledby_name(element)
+      yielder << aria_label_name(element)
+      yielder << referenced_label_name(element)
+      yielder << ancestor_label_text_name(element)
+    end.lazy.compact.first
```

Comment:
As fun as this was to implement, maybe it's no better than just a set of `||` fallbacks and taking advantage of short-circuit boolean evaluation?

```suggestion
    hidden_name(element) ||
      aria_labelledby_name(element) ||
      aria_label_name(element) ||
      referenced_label_name(element) ||
      ancestor_label_text_name(element)
```